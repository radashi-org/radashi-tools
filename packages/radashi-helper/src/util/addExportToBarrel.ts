/**
 * This function assumes the `content` represents a barrel file, which
 * simply re-exports a list of modules. It also assumes you want to
 * export everything from the given `specifier`. Lastly, it assumes
 * the barrel is alphabetically sorted and all lines begin with
 * `export * from` and use single quotes for the specifiers.
 *
 * In other words, this function is brittle and purpose-built.
 */
export function addExportToBarrel(content: string, specifier: string) {
  const addedLine = `export * from '${specifier}'`
  const existingLines = content.split('\n')

  let insertIndex = Math.max(existingLines.length - 1, 0)

  for (let i = 0; i < existingLines.length; i++) {
    if (existingLines[i] && existingLines[i] > addedLine) {
      if (i === 0) {
        insertIndex = 0
        break
      }
      for (let j = i - 1; j >= 0; j--) {
        if (existingLines[j].length) {
          insertIndex = j + 1
          break
        }
      }
      break
    }
  }

  existingLines.splice(insertIndex, 0, addedLine)

  return existingLines.join('\n')
}
