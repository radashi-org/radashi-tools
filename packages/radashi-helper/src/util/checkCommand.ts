import { exec } from 'node:child_process'

export async function checkCommand(cmd: string) {
  return new Promise<boolean>(resolve => {
    exec(`command -v ${cmd}`, error => {
      resolve(!error)
    })
  })
}
