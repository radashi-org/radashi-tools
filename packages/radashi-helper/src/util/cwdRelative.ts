import { white } from 'kleur/colors'
import { relative } from 'node:path'

export function cwdRelative(path: string) {
  let result = relative(process.cwd(), path)
  if (!result.startsWith('..')) {
    result = `./${result}`
  }
  return white(result)
}
