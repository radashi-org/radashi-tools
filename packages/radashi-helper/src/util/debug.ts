import createDebug from 'debug'

export const debug = createDebug('radashi')
