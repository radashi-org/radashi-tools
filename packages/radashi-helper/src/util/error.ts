import { isError } from 'radashi'
import { log } from './log'

export class RadashiError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'RadashiError'
  }
}

export class EarlyExitError extends RadashiError {
  constructor(message = 'Cannot continue. Exiting...') {
    super(message)
    this.name = 'EarlyExitError'
  }
}

export function forwardStderrAndRethrow(error: any): never {
  if (isError(error) && 'stderr' in error) {
    if ('stdout' in error) {
      log(error.stdout as string)
    }
    log.error(error.stderr as string)
  }
  throw error
}
