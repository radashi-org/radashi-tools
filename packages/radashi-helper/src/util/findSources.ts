import glob from 'fast-glob'
import type { Env } from '../env'

const fileTypes = ['src', 'overrides', 'rewired'] as const

type FileType = (typeof fileTypes)[number]

export async function findSources<Only extends FileType = FileType>(
  env: Env,
  only?: Only[],
): Promise<{
  src: 'src' extends Only ? string[] : undefined
  overrides: 'overrides' extends Only ? string[] : undefined
  rewired: 'rewired' extends Only ? string[] : undefined
}>

export async function findSources(
  env: Env,
  only?: FileType[],
): Promise<{
  src: string[] | undefined
  overrides: string[] | undefined
  rewired: string[] | undefined
}> {
  const [src, overrides, rewired] = await Promise.all(
    [
      {
        id: 'src' as const,
        glob: ['src/**/*.ts', '!src/*.ts'],
      },
      {
        id: 'overrides' as const,
        glob: ['overrides/src/**/*.ts', '!overrides/src/*.ts'],
      },
      {
        id: 'rewired' as const,
        glob: ['overrides/rewired/**/*.ts'],
      },
    ].map(file =>
      only?.includes(file.id) !== false
        ? glob(file.glob, {
            cwd: env.root,
            absolute: true,
          })
        : undefined,
    ),
  )

  return {
    src,
    overrides,
    rewired,
  }
}
