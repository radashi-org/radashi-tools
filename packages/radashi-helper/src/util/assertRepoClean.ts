import { RadashiError } from './error'
import { isRepoClean } from './isRepoClean'

export async function assertRepoClean(cwd: string) {
  if (!(await isRepoClean(cwd))) {
    throw new RadashiError(
      'Your repository has uncommitted changes.' +
        ' Please commit or stash them before overriding.',
    )
  }
}
