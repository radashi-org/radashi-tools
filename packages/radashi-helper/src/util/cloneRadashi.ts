import { execa, type Options as ExecaOptions } from 'execa'
import type { PackageJson } from 'type-fest'
import { forwardStderrAndRethrow, RadashiError } from './error'
import { prepareGitHubDefaultRepo } from './github'

export async function getInstalledRadashiRef(pkg: PackageJson) {
  const specifier = pkg.dependencies?.radashi
  if (!specifier) {
    throw new RadashiError('No radashi dependency found in package.json')
  }

  let ref = specifier.split('@').pop()!

  if (ref === 'beta') {
    ref = 'main'
  } else if (ref !== 'next') {
    ref = await execa('npm', ['view', 'radashi@' + ref, '--json']).then(
      ({ stdout }) => 'v' + JSON.parse(stdout).version,
    )
  }

  return ref
}

export function isPreReleaseRadashiTag(ref: string) {
  return /-(beta|alpha)/.test(ref)
}

export function getRadashiCloneURL(ref: string) {
  let cloneUrl = 'https://github.com/radashi-org/radashi'
  if (isPreReleaseRadashiTag(ref)) {
    // The "radashi-beta" repo is where versioned pre-release tags are
    // stored, which helps to avoid spamming the main repo.
    cloneUrl += '-beta'
  }
  return cloneUrl
}

export async function cloneRadashi(
  ref: string,
  dir: string,
  opts: ExecaOptions = {},
) {
  const cloneUrl = getRadashiCloneURL(ref)
  await execa(
    'git',
    ['clone', cloneUrl, '--branch', ref, '--single-branch', dir],
    opts,
  ).catch(error => {
    if (opts.stdio == null) {
      forwardStderrAndRethrow(error)
    }
    throw error
  })
  await prepareGitHubDefaultRepo(dir)
}
