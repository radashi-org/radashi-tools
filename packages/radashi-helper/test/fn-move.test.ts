import {
  expectGitDiffToMatchSnapshot,
  expectLogsToMatchSnapshot,
  expectOpenedFilesToMatchSnapshot,
  expectPromptsToMatchSnapshot,
} from './util/expect'
import { prompt } from './util/mocks'
import { run } from './util/run'

test('move a function', async () => {
  const responses: Record<string, any> = {
    selectedFunc: 'number/add',
    action: 'both',
    selectedGroup: 'new',
    newGroup: 'test',
    newFuncName: 'combine',
  }

  prompt.mockImplementation(async options => {
    return responses[options.name]
  })

  await run('fn move')

  expectLogsToMatchSnapshot()
  expectPromptsToMatchSnapshot()
  expectOpenedFilesToMatchSnapshot()
  await expectGitDiffToMatchSnapshot('all-staged')
})
