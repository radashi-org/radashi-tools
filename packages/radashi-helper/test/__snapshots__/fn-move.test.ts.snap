// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`move a function > git diff (all-staged) 1`] = `
"diff --git a/benchmarks/number/add.bench.ts b/benchmarks/test/combine.bench.ts
similarity index 100%
rename from benchmarks/number/add.bench.ts
rename to benchmarks/test/combine.bench.ts
diff --git a/docs/number/add.mdx b/docs/test/combine.mdx
similarity index 100%
rename from docs/number/add.mdx
rename to docs/test/combine.mdx
diff --git a/src/number/add.ts b/src/test/combine.ts
similarity index 100%
rename from src/number/add.ts
rename to src/test/combine.ts
diff --git a/tests/number/add.test.ts b/tests/test/combine.test.ts
similarity index 100%
rename from tests/number/add.test.ts
rename to tests/test/combine.test.ts"
`;

exports[`move a function > interactive prompts 1`] = `
[
  [
    {
      "choices": [
        {
          "title": "number/add",
          "value": "number/add",
        },
      ],
      "message": "Select a function to move:",
      "name": "selectedFunc",
      "type": "autocomplete",
    },
  ],
  [
    {
      "choices": [
        {
          "title": "Move function to a new group",
          "value": "move",
        },
        {
          "title": "Rename function",
          "value": "rename",
        },
        {
          "title": "Both move and rename",
          "value": "both",
        },
      ],
      "message": "What would you like to do?",
      "name": "action",
      "type": "select",
    },
  ],
  [
    {
      "choices": [
        {
          "title": "Create a new group",
          "value": "new",
        },
        {
          "title": "array",
          "value": "array",
        },
        {
          "title": "async",
          "value": "async",
        },
        {
          "title": "curry",
          "value": "curry",
        },
        {
          "title": "function",
          "value": "function",
        },
        {
          "title": "number",
          "value": "number",
        },
        {
          "title": "object",
          "value": "object",
        },
        {
          "title": "random",
          "value": "random",
        },
        {
          "title": "series",
          "value": "series",
        },
        {
          "title": "string",
          "value": "string",
        },
        {
          "title": "typed",
          "value": "typed",
        },
      ],
      "message": "Select a group for the function:",
      "name": "selectedGroup",
      "type": "autocomplete",
    },
  ],
  [
    {
      "message": "Enter the new group name:",
      "name": "newGroup",
      "type": "text",
    },
  ],
  [
    {
      "message": "Enter the new function name:",
      "name": "newFuncName",
      "type": "text",
    },
  ],
]
`;

exports[`move a function > logs 1`] = `
"[info] Renaming ./src/number/add.ts to ./src/test/combine.ts
[info] Removing empty directory: ./src/number
[info] Renaming ./docs/number/add.mdx to ./docs/test/combine.mdx
[info] Removing empty directory: ./docs/number
[info] Renaming ./benchmarks/number/add.bench.ts to ./benchmarks/test/combine.bench.ts
[info] Removing empty directory: ./benchmarks/number
[info] Renaming ./tests/number/add.test.ts to ./tests/test/combine.test.ts
[info] Removing empty directory: ./tests/number"
`;

exports[`move a function > opened files 1`] = `[]`;
